hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/QuasiClosedContainerHandler.java

```diff
@@ -0,0 +1,130 @@
+  }
+
+  /**
+   * If possible, force closes the Ratis container in QUASI_CLOSED state.
+   * Replicas with the highest Sequence ID are selected to be closed.
+   * @param request ContainerCheckRequest object representing the container
+   * @return true if close commands were sent, otherwise false
+   */
+  @Override
+  public boolean handle(ContainerCheckRequest request) {
+    ContainerInfo containerInfo = request.getContainerInfo();
+    if (containerInfo.getReplicationType() == HddsProtos.ReplicationType.EC) {
+      return false;
+    }
+
+    if (containerInfo.getState() != HddsProtos.LifeCycleState.QUASI_CLOSED) {
+      return false;
+    }
+
+    Set<ContainerReplica> replicas = request.getContainerReplicas();
+    if (canForceCloseContainer(containerInfo, replicas)) {
+      forceCloseContainer(containerInfo, replicas);
+      return true;
+    } else {
+      request.getReport().incrementAndSample(
+          ReplicationManagerReport.HealthState.QUASI_CLOSED_STUCK,
+          containerInfo.containerID());
+    }
+    return false;
+  }
+
+  /**
+   * Returns true if more than 50% of the container replicas with unique
+   * originNodeId are in QUASI_CLOSED state.
+   *
+   * @param container Container to check
+   * @param replicas Set of ContainerReplicas
+   * @return true if we can force close the container, false otherwise
+   */
+  private boolean canForceCloseContainer(final ContainerInfo container,
+      final Set<ContainerReplica> replicas) {
+    final int replicationFactor =
+        container.getReplicationConfig().getRequiredNodes();
+    final long uniqueQuasiClosedReplicaCount = replicas.stream()
+        .filter(r -> r.getState() == State.QUASI_CLOSED)
+        .map(ContainerReplica::getOriginDatanodeId)
+        .distinct()
+        .count();
+    return uniqueQuasiClosedReplicaCount > (replicationFactor / 2);
+  }
+
+  /**
+   * Force close the container replica(s) with the highest Sequence ID.
+   *
+   * <p>
+   *   Note: We should force close the container only if >50% (quorum)
+   *   of replicas with unique originNodeId are in QUASI_CLOSED state.
+   * </p>
+   *
+   * @param container ContainerInfo
+   * @param replicas Set of ContainerReplicas
+   */
+  private void forceCloseContainer(final ContainerInfo container,
+      final Set<ContainerReplica> replicas) {
+    final List<ContainerReplica> quasiClosedReplicas = replicas.stream()
+        .filter(r -> r.getState() == State.QUASI_CLOSED)
+        .collect(Collectors.toList());
+
+    final Long sequenceId = quasiClosedReplicas.stream()
+        .map(ContainerReplica::getSequenceId)
+        .max(Long::compare)
+        .orElse(-1L);
+
+    LOG.info("Force closing container {} with BCSID {}, which is in " +
+            "QUASI_CLOSED state.", container.containerID(), sequenceId);
+
+    quasiClosedReplicas.stream()
+        .filter(r -> sequenceId != -1L)
+        .filter(replica -> replica.getSequenceId().equals(sequenceId))
```

Comment:
I don't think we have a unit test to validate we only send close commands for Replicas matching the max sequence number. It would be good to add one, as this is an important piece of logic I think.